### PR TITLE
feat: enhance scrollIntoView to support native mobile

### PR DESCRIPTION
**Pull request by wswebcreation**
Initial setup to support Native apps

## Proposed changes

This PR enhances the `scrollIntoView` to support native mobile apps by using the native `mobile:execute` methods of the drivers

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added the necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by wswebcreation** [comment]
Hmm, getting the following error when I try to test the command

```logs
Can't call scrollIntoView on element with selector "//android.widget.TextView[@resource-id="android:id/title" and @text="System"]" because element wasn't found
```

For native the elements are not always available as with web. I might have found a hack I'm not proud of. The [`implicitWait`](https://github.com/webdriverio/webdriverio/blob/main/packages/webdriverio/src/utils/implicitWait.ts#L11) excludes checking for the `*waitUntil*` methods. If I just "hack" the `scrollIntoView` into it, it works, it will not find the element, it will scroll and check if it exists, if not, it will scroll, and so on. 
This is a "hacky" solution, so I would love to get some input/look into a specific direction to make this more stable. I'm open to suggestions 😅

File: packages/webdriverio/src/commands/element/scrollIntoView.ts [hunk 1]
```diff
@@ -2,9 +2,127 @@ import logger from '@wdio/logger'
 import { ELEMENT_KEY } from 'webdriver'
 
 import { getBrowserObject } from '@wdio/utils'
+import type { CustomScrollIntoViewOptions, MobileScrollIntoViewOptions } from 'src/types.js'
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/scrollIntoView.ts:5]:
```suggestion
import type { CustomScrollIntoViewOptions, MobileScrollIntoViewOptions } from '../../types.js'
```

File: packages/webdriverio/src/commands/element/scrollIntoView.ts [hunk 2]
```diff
@@ -2,9 +2,127 @@ import logger from '@wdio/logger'
...
 
 const log = logger('webdriverio')
 
+type MobileScrollUntilVisibleOptions = {
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/scrollIntoView.ts:9]: Can we move the code down, below the command function? Having other types or functions defined before just invites to add comments to them which will ultimately replace the command documentation when generating the docs

File: packages/webdriverio/src/commands/element/scrollIntoView.ts [hunk 3]
```diff
@@ -2,9 +2,127 @@ import logger from '@wdio/logger'
...
+    const scrollableElements = (await browser.$$(
+        selector
+    )) as unknown as WebdriverIO.Element[]
+
+    if (scrollableElements.length > 0) {
+        return scrollableElements[0]
+    }
+
+    throw new Error(
+        `Default scrollable element '${browser.isIOS ? defaultIosSelector : defaultAndroidSelector}' not found.`
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/scrollIntoView.ts:44]: What should the user do in this case? Is this a common scenario? If so, we should add a recommendation on what a user should try next.

File: packages/webdriverio/src/commands/element/scrollIntoView.ts [hunk 4]
```diff
@@ -2,9 +2,127 @@ import logger from '@wdio/logger'
...
+    element,
+    options
+}: {
+    browser: WebdriverIO.Browser,
+    element: WebdriverIO.Element,
+    options: MobileScrollIntoViewOptions
+    }) {
+    const defaultOptions = {
+        maxScrolls: 10,
+        scrollDirection: MobileScrollDirection.Down,
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/scrollIntoView.ts:100]: What if we already scrolled all the way down and now have to do scroll up to find the element?

**Comment by christian-bromann** [comment]
> If I just "hack" the `scrollIntoView` into it, it works, it will not find the element, it will scroll and check if it exists, if not, it will scroll, and so on.

It seem fine to me if we extend the if statement with scrollIntoView if we are in a mobile and native context, e.g.

```ts
const browser = getBrowserObject(currentElement)
const shallSkipForMobileScrollIntoView = browser.isMobile && (await browser.getContext === 'NATIVE')
```

File: packages/webdriverio/src/commands/element/scrollIntoView.ts (hunk 1 above)
Comment (wswebcreation) [reply packages/webdriverio/src/commands/element/scrollIntoView.ts]: Thanks, accepted it

File: packages/webdriverio/src/commands/element/scrollIntoView.ts (hunk 2 above)
Comment (wswebcreation) [reply packages/webdriverio/src/commands/element/scrollIntoView.ts:9]: sure, I added it to the top due to coding conventions that we first need to declare before we use

File: packages/webdriverio/src/commands/element/scrollIntoView.ts (hunk 3 above)
Comment (wswebcreation) [reply packages/webdriverio/src/commands/element/scrollIntoView.ts:44]: Good suggestion, I'll update the error

File: packages/webdriverio/src/commands/element/scrollIntoView.ts (hunk 4 above)
Comment (wswebcreation) [reply packages/webdriverio/src/commands/element/scrollIntoView.ts:100]: I'll update the error message if we can't find an element. I believe that users need to know their app in which direction to swipe to find the element. Sadly, native app automation is not that easy in comparison to web 😅

**Comment by wswebcreation** [comment]
> > If I just "hack" the `scrollIntoView` into it, it works, it will not find the element, it will scroll and check if it exists, if not, it will scroll, and so on.
> 
> It seem fine to me if we extend the if statement with scrollIntoView if we are in a mobile and native context, e.g.
> 
> ```ts
> const browser = getBrowserObject(currentElement)
> const shallSkipForMobileScrollIntoView = browser.isMobile && (await browser.getContext === 'NATIVE')
> ```

Thanks, that's a solution. If we have the `isNativeContext` we could even determine better. There's only one downside, and that is that will always do an extra webdriver call for mobile, no matter if it's web or native app

**Comment by christian-bromann** [comment]
> There's only one downside, and that is that will always do an extra webdriver call for mobile, no matter if it's web or native app

How about enhancing the context manager the way we discussed so the extra call is obsolete?

**Comment by wswebcreation** [comment]
> > There's only one downside, and that is that will always do an extra webdriver call for mobile, no matter if it's web or native app
> 
> 
> 
> How about enhancing the context manager the way we discussed so the extra call is obsolete?

That's also what I wanted to do, but in a separate PR which would "revert" the extra call. I wanted to create "smaller" PRs because I don't have a lot of time the coming weeks

**Comment by wswebcreation** [comment]
@christian-bromann 

Not sure why other UTs are failing, I can't see any impact due to my changes (to the fetch Mock). Any ideas?

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

**Pull request merged**